### PR TITLE
scripts: Store and delete mem for deferred ops

### PR DESCRIPTION
**Pull request by larsivarsimonsen-arm**
Adds tracking of allocations associated with deferred operations with
subsequent deletion once the deferred operations have completed.

Closes KhronosGroup#2767

### CODE REVIEW

**Review by ncesario-lunarg — COMMENTED** [review]
This generally LGTM.

One question I have is why does `deferred_operation_cleanup` have to map to a `std::function`, and the call that function to delete the backing data? At first glance it seems like it would be possible for the value to be the actual `safe_Vk*` struct (could be multiple maps, or a union of all possible maps), using the map entry as the backing data, and then just erase the entry at `DispatchDeferredOperationJoinKHR` time. If I understand this correctly, I realize this change is "less intrusive," but it would be nice to see _less_ "raw heap allocations" than adding new ones.

**Comment by jeremyg-lunarg** [comment]
> One question I have is why does `deferred_operation_cleanup` have to map to a `std::function`, and then call that function to delete the backing data? At first glance it seems like it would be possible for the value to be the actual `safe_Vk*` struct (could be multiple maps, or a union of all possible `safe_Vk*` types), using the map entry as the backing data, and then just erase the entry at `DispatchDeferredOperationJoinKHR` time. If I understand this correctly, I realize this change is "less intrusive," but it would be nice to see _less_ "raw heap allocations" than adding new ones.

This part was my suggestion, because I was concerned we'd hit a case where we had to free more than 1 safe_ object at a time and that we'd want this flexibility. Also trying to make a union in the scripted code seemed tricky.  But if we can find a way to make it simpler that'd be fine w/me. Multiple maps probably wouldn't be good because then we'd have to do multiple lookups when the deferred operation completes.   It *might* also be possible to use unique_ptr here.

File: layers/generated/layer_chassis_dispatch.cpp [hunk 1]
```diff
@@ -1247,6 +1247,31 @@ VkResult DispatchBeginCommandBuffer(VkCommandBuffer commandBuffer, const VkComma
...
+    VkResult result = layer_data->device_dispatch_table.DeferredOperationJoinKHR(device, operation);
+
+    // If this thread completed the operation, free any retained memory.
+    if (result == VK_SUCCESS)
+    {
+        auto iter = layer_data->deferred_operation_cleanup.pop(operation);
+        if (iter != layer_data->deferred_operation_cleanup.end())
+        {
+            std::function<void()> &delete_fn = iter->second;
+            delete_fn();
```
[the remark this answers is not in the document]
Comment (larsivarsimonsen-arm) [reply layers/generated/layer_chassis_dispatch.cpp:1268]: I think it's more clear to explicitly show it's a function, but calling `iter->second()` would also work.

**Comment by larsivarsimonsen-arm** [comment]
> [...]because I was concerned we'd hit a case where we had to free more than 1 safe_ object at a time and that we'd want this flexibility.

For now, this is not the case, but I suppose it's possible we might hit this in the future.

> Also trying to make a union in the scripted code seemed tricky.  But if we can find a way to make it simpler that'd be fine w/me.

I think we would need to go through all functions that support deferred operations, look up all safe_-structs for those and add them to the union. It should be doable but tricky.

>  It might also be possible to use unique_ptr here.

Yes, I considered that and it should work, but I think we would need to make a `unique_ptr<void>` and explicitly supply the destructor for each case. I didn't like the way that looked, so I went with lambda functions instead.

File: layers/generated/layer_chassis_dispatch.cpp [hunk 2]
```diff
@@ -9493,11 +9525,16 @@ VkResult DispatchCreateRayTracingPipelinesKHR(
...
+        if (deferredOperation != VK_NULL_HANDLE) {
+            layer_data->deferred_operation_cleanup.insert(deferredOperation, [local_pCreateInfos](){ delete[] local_pCreateInfos; });
+        } else {
+            delete[] local_pCreateInfos;
+        }
     }
     {
         for (uint32_t index0 = 0; index0 < createInfoCount; index0++) {
-            if (pPipelines[index0] != VK_NULL_HANDLE) {
+            if (pPipelines[index0] != VK_NULL_HANDLE && deferredOperation != VK_NULL_HANDLE) {
```
Comment (larsivarsimonsen-arm) [inline layers/generated/layer_chassis_dispatch.cpp:9536]: A bug here: should be `deferredOperation == VK_NULL_HANDLE`

Comment (AaronHaganAMD) [reply layers/generated/layer_chassis_dispatch.cpp]: Should the pipeline get wrapped after the final DispatchDeferredOperationJoinKHR. After the pipeline is created, calling  DispatchGetRayTracingShaderGroupHandlesKHR..  pipeline = layer_data->Unwrap(pipeline); is resulting in null pipeline.

File: layers/generated/layer_chassis_dispatch.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ncesario-lunarg) [reply layers/generated/layer_chassis_dispatch.cpp:1268]: If you feel that's clearer, that's good with me.

**Review by ncesario-lunarg — APPROVED** [review]
@jeremyg-lunarg @larsivarsimonsen-arm I think I'm convinced this should stay as is now :)

I would still like to see the raw `new`s get replaced at some point, but that can be part of a future "cleanup/refactor."

If there is a likely need for future custom logic at cleanup time, one option might be to create a struct that stores the data, and put that logic into the destructor. A lambda storing a heap-allocated pointer just kind of gives me the opposite of a "warm and fuzzy feeling."

**Comment by AaronHaganAMD** [comment]
I believe the pipeline needs wrapped at the end of the deferred pipeline build otherwise it will end up as a null when unwrapped in following commands.

File: layers/generated/layer_chassis_dispatch.cpp (hunk 2 above)
Comment (larsivarsimonsen-arm) [reply layers/generated/layer_chassis_dispatch.cpp]: That's a good point.

I think it should be possible to do the wrapping in the cleanup lambda function by capturing pPipelines, createInfoCount and potentially layer_data (can alternatively be passed as a parameter in DispatchDeferredOperationJoinKHR).

WrapNew appears to be thread safe, as unique_id_mapping is a vl_concurrent_unordered_map.

**Comment by larsivarsimonsen-arm** [comment]
I've updated the fix with WrapNew in the lambda function for DispatchCreateRayTracingPipelinesKHR.
@AaronHaganAMD: Would you be able to test this for me? I currently don't have access to a device to test it on.

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM. Looks like the need for arbitrary cleanup code in the lambda showed up sooner than we thought. :)

**Comment by AaronHaganAMD** [comment]
> I've updated the fix with WrapNew in the lambda function for DispatchCreateRayTracingPipelinesKHR. @AaronHaganAMD: Would you be able to test this for me? I currently don't have access to a device to test it on.

PostCallRecordCreateRayTracingPipelinesKHR needs to be called from the lambda function somehow where the pipeline handle is wrapped, so it gets stored in the object tracking...

**Pull request merged**